lib/winston-graylog2.js

```diff
@@ -79,12 +80,22 @@ let Graylog2 = (winston.transports.Graylog2 = function(options) {
   this.name = options.name || 'graylog2';
   this.silent = options.silent || false;
   this.handleExceptions = options.handleExceptions || false;
-  this.prelog =
-    typeof options.prelog === 'function' ? options.prelog : _.identity;
+  this.json = options.json || false;
+  this.colorize = options.colorize || false;
+  this.label = options.label || null;
+  this.prettyPrint = options.prettyPrint || false;
+  this.timestamp = options.timestamp || true;
+  this.logstash = options.logstash || null;
+  this.depth = options.depth || null;
+  this.showLevel = options.showLevel === undefined ? true : options.showLevel;
```

Comment:
Hi @surbhi-mahajan. Thanks for your contribution!

I have a question regarding these changes. Why do we need all these options (e.g. `json`, `colorize`, `logstash`) in a transport for graylog? How does it relate to the action of sending messages into graylog? Looks more like a copy-paste from Console transport https://github.com/winstonjs/winston/blob/2.4.0/lib/winston/transports/console.js#L25
I don't see how it fixes the issue with passing an error object as the first argument (instead of a string message) to `#error` method.

I may be wrong, so please explain a bit more why exactly we need this code here.

Reply:
@schfkt In this PR, I have added 2 things: bugfix and enhancement supporting various formatters.
Bug fix:  My use case is I want to add line number, function name at which error has occurred in a log message field. So, to include the line number, function name in my log message field, prelog function should be called with error stack and error message. Moreover, Winston provides error object to winstron-graylog2 in the form of meta object if #log.error method is called with error object but prelog function is only called with message i.e empty string.

**Format** of log which I need in graylog are as follow:

```
full_message
{"message":"<LEVEL>   <TIMESTAMP>  <FILENAME>  <METHODNAME><LINENUMBER> : <MESSAGE>}
level
3
timestamp
2018-08-28T07:59:21.266Z
```

Enchancement : Second thing I have added is usage of common.log method provided by Winston library. It supports message in various formats like json, raw etc and moreover, it automatically calls formatter method with options including meta (error stack and message in case of log.error(new Errror('here'))), message etc. I have taken reference from built-in winston tranport.

Reply:
@surbhi-mahajan 

> Moreover, Winston provides error object to winstron-graylog2 in the form of meta object if #log.error method is called with error object but prelog function is only called with message i.e empty string which is incorrect message

> So, we need to call prelog function with error stack to get above format.

Why didn't you just improved the usage of `prelog` option to do that?

Anyway, my point is that all these options are specific to only Console and File transports. For example: colorize. How will it affect messages sent into graylog? AFAIK, graylog doesn't support this.

If you want to make it possible to preprocess messages and metadata -- improve existing logic, which leverages `prelog` option, instead of just copying some things from the other transport (which are specific to only that transport). 
https://github.com/namshi/winston-graylog2/blob/master/lib/winston-graylog2.js#L99 there are everything for that: level, message and meta parameters.

Reply:
@surbhi-mahajan that way is not only simpler, but it won't break backward-compatibility of the library (i.e. you removed `prelog` completely).

Reply:
I agreed graylog does not support colorize, prettyPrint but it supports all other options like json, logstash, raw, timestamp, stringify, custom formatter (i.e prelog in this case) etc. My point is by using common.log method of Winston, it automatically supports above formats and common.log method also passes meta ,message and other options to custom formatter (i.e prelog option) for us. In this way, user does not need to provide his own custom formatter (i.e prelog option) in case he wants any above formats (json, raw, logstash etc).

Yes, above approach that you have mentioned is simpler and it also preprocess meta and message both But I think it will be great if we reuse common.log method provided by Winston as it automatically supports various formatters and pass meta data to custom formatters for us. Moreover, another benefit is code consistency between various custom and built-in transports as formatter option is also used in built-in transports.

Reply:
@surbhi-mahajan 

> timestamp

Why do you want to store a timestamp in the `message` field in graylog? Do you know that graylog assigns a separate `timestamp` field for each message it gets? For example:

![timestamp](https://user-images.githubusercontent.com/3599143/44835182-095e7280-ac3c-11e8-9b10-300f17f5f4e2.png)

And you end up with two different timestamps.